sympy/sets/sets.py

```diff
@@ -1202,28 +1202,32 @@ def reduce(args):
 
         # ===== Pair-wise Rules =====
         # Here we depend on rules built into the constituent sets
-        args = set(args)
+        args = FiniteSet(*args)
```

Comment:
I thought of providing an alternate implementation of this part myself but I'm running to strange errors related to types. Will try again tomorrow morning.